HIP/hip-513.md

```diff
@@ -0,0 +1,293 @@
+    /**
+     * Bytes passed in as input data to this action.
+     */
+    bytes input = 5;
+
+    /**
+     * Who this action is directed to.
+     */
+    oneof recipient {
+        /**
+         * The AccountID of the recipient if the recipient is an account. Only HBars will be transferred, no other side
+         * effects should be expected.
+         */
+        AccountID recipientAccount = 6;
+
+        /**
+         * The ContractID of the recipient if the recipient is a smart contract.
+         */
+        ContractID recipientContract = 7;
+
+        /**
+         * If the action was directed at an invalid solidity address, what that address was.
+         */
+        bytes invalidSolidityAddress = 8;
+    }
+
+    /**
+     * The value (int tinybars) that is associated with this action.
+     */
+    int64 value = 9;
+
+    /**
+     * The actual gas spent by this action.
+     */
+    uint64 gasUsed = 10;
+
+    /**
+     * Was the action successful?
+     */
+    bool success = 11;
+
+    /**
+     * The result data of the action.
+     */
+    oneof resultData {
+
+        /**
+         * If successful, the output bytes of the action.
+         */
+        bytes output = 12;
+
+        /**
+         * The contract itself cause the transaction to fail via the `REVERT` operation
+         */
+        bytes revertReason = 13;
+
+        /**
+         * The transaction itself failed without an explicit `REVERT`
+         */
+        bytes error = 14;
+    }
+
+    /**
+     * The nesting depth of this call. The original action is at depth=0.
+     */
+    uint32 callDepth = 15;
+}
+```
+
+For every smart contract transaction a `ContractAction` record should be added to the sidecar record.
+
+### Add ContractBytecode sidecar record type
+
+Add and populate a new sidecar record of type `ContractBytecode`. It will
+have two fields - `deploymentBytecode` and `runtimeBytecode`. Fields will
+allow tracking smart contracts.
+
+```
+message ContractBytecode {
+
+    /**
+    * Contract bytecode during deployment
+    */
+    bytes deploymentBytecode = 1;
+    
+    /**
+    * Contract bytecode after deployment
+    */
+    bytes runtimeBytecode = 2;
+}
+```
+
+For every contract create `ContractBytecode` should be added to the sidecar record.
+
+### Remove ContractStateChange from ContractFunctionResult
+
+ContractStateChange will be removed from ContractFunctionResult. It will be part of sidecar transaction record.
+
+```
+// existing protobuf
+message ContractFunctionResult {
+    // existing fields
+    ContractID contractID = 1;
+    bytes contractCallResult = 2;
+    string errorMessage = 3;
+    bytes bloom = 4;
+    uint64 gasUsed = 5;
+    repeated ContractLoginfo logInfo = 6;
+    repeated ContractID createdContractIDs = 7;
+    
+    // field to remove
+    repeated ContractStateChange stateChanges = 8;
+}
+```
+
+## Backwards Compatibility
+
+This HIP changes ContractFunctionResult protobuf by removing ContractStateChange from it.
```

Comment:
We should discuss the impact of removing a proto field.  In the event it was populated it will be ignored by future processing, but will not break the validity of mesasges containing it..  IIRC previewnet and testnet had this enabled.